sympy/combinatorics/rewritingsystem.py

```diff
@@ -0,0 +1,206 @@
+            s2 = s2*g**-1
+            if len(s1) - len(s2) in [0, 1, 2]:
+                if s2 < s1:
+                    self.rules[s1] = s2
+                    new_keys.add(s1)
+                elif s1 < s2:
+                    self.rules[s2] = s1
+                    new_keys.add(s2)
+
+        # overlaps on the left
+        while len(w1) - len(w2) > 1:
+            g = w1[0]
+            w1 = w1.subword(1, len(w1))
+            w2 = g**-1*w2
+            if len(w1) - len(w2) in [0, 1, 2]:
+                if w2 < w1:
+                    self.rules[w1] = w2
+                    new_keys.add(w1)
+                elif w1 < w2:
+                    self.rules[w2] = w1
+                    new_keys.add(w2)
+
+        return new_keys
+
+    def _remove_redundancies(self):
+        '''
+        Reduce left- and right-hand sides of reduction rules
+        and remove redundant equations (i.e. those for which
+        lhs == rhs)
+
+        '''
+        rules = self.rules.copy()
+        for r in rules:
+            v = self.reduce(r, exclude=r)
+            w = self.reduce(rules[r])
+            if v != r:
+                del self.rules[r]
+                if v != w:
+                    self.add_rule(v, w)
+            else:
+                self.rules[v] = w
+        return
+
+    def make_confluent(self, check=False):
+        '''
+        Try to make the system confluent using the Knuth-Bendix
+        completion algorithm
+
+        '''
+        lhs = list(self.rules.keys())
+
+        def _overlaps(r1, r2):
+            len1 = len(r1)
+            len2 = len(r2)
+            result = []
+            j = 0
+            while j < len1 + len2 - 1:
+                j += 1
+                if (r1.subword(len1 - j, len1 + len2 - j, strict=False)
+                       == r2.subword(j - len1, j, strict=False)):
+                    a = r1.subword(0, len1-j, strict=False)
+                    a = a*r2.subword(0, j-len1, strict=False)
+                    b = r2.subword(j-len1, j, strict=False)
+                    c = r2.subword(j, len2, strict=False)
+                    c = c*r1.subword(len1 + len2 - j, len1, strict=False)
+                    result.append(a*b*c)
+            return result
+
+        def _process_overlap(w, r1, r2):
+                s = w.eliminate_word(r1, self.rules[r1])
+                t = w.eliminate_word(r2, self.rules[r2])
+                if self.reduce(s) != self.reduce(t):
+                    new_keys = self.add_rule(t, s)
+                    return new_keys
+                return
+
+        added = 0
+        for i, r1 in enumerate(lhs):
+            # this could be lhs[i+1:] to not
+            # check each pair twice but lhs
+            # is extended in the loop and the new
+            # elements have to be checked with the
+            # preceding ones. there is probably a better way
+            # to handle this
+            for r2 in lhs:
+                overlaps = _overlaps(r1, r2)
```

Comment:
Is it necessary to do this also when r1 and r2 are the same? It seems that, in that case, the overlap w is also the same, and then `_process_overlap(w, r1, r2)` would do nothing.